### ISSUE TITLE
[question] Propagate header-only dependencies downstream

### ISSUE THREAD

**Issue by hobbeshunter** [conan-io/conan#13538]
### What is your question?

I'm trying to package [json2cpp](https://github.com/lefticus/json2cpp). It consists of an executable (a potential user might want to call via CMake) and a header-only library. 

This header-only library has a dependency on another header-only library: valijson.

Right know I cannot get this transitive dependency propagated downstream.

What I've done so far can be seen at https://github.com/hobbeshunter/json2cpp/tree/9145ad94d0b725eaa4599a9a7770e0ecc79d2cf6

As can be seen in the test package, I need to add valijson as dependency to any consumer:
https://github.com/hobbeshunter/json2cpp/blob/3b27544cdb73fa7692f14cef27701d3147f4f17d/test_package/conanfile.py#L12

To consume the tool and the library from json2cpp I set the traits `headers`, `run` and `build`: https://github.com/hobbeshunter/json2cpp/blob/3b27544cdb73fa7692f14cef27701d3147f4f17d/test_package/conanfile.py#LL13C4-L13C4

Without declaring valijson as dependency in the test package the `valijson-*.cmake` files are not generated by the CMakeDeps generator and so the valijson package can't be found by cmake.

Does anybody know how I can propagate the dependency downstream?

I've seen that other people have similar issues: https://github.com/conan-io/conan/pull/12970. But as to them having tool and library in different packages doesn't seem like a good solution as those two are tightly coupled.

conan: 2.0.2
OS: Ubuntu 20.04
compiler: gcc-9 and gcc-10
CMake: 3.26.1

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @hobbeshunter 

Thanks for your question.
I am having a look at https://github.com/hobbeshunter/json2cpp/tree/9145ad94d0b725eaa4599a9a7770e0ecc79d2cf6, but I only find there a ``conanfile.txt``. A ``conanfile.txt`` is not good to create a package, but a ``conanfile.py`` is needed.

Then I see the ``test_package``, but doesn't seem to belong to the same commit? I am a bit confused.

Some hints:
- The extra ``requires()`` to the transitive dependency ``valijson`` should indeed not be necessary in consumers
- The trait ``transitive_headers=True`` should be set in the ``self.requires("valijson/...", transitive_headers=True)`` that will be in the ``conanfile.py`` of json2cpp

**Comment by hobbeshunter** [comment]
Hi @memsharded 
thank you very much for your answer. Im very sorry, I mixed up some links. The first one was wrong, here is the correct one: https://github.com/hobbeshunter/json2cpp/tree/3b27544cdb73fa7692f14cef27701d3147f4f17d

There you can see, that I replaced the old `conanfile.txt` of json2cpp with a `conanfile.py`.  And there I have set `transitive_headers`  on valijson: https://github.com/hobbeshunter/json2cpp/blob/3b27544cdb73fa7692f14cef27701d3147f4f17d/conanfile.py#L29

**Comment by hobbeshunter** [comment]
The problem seems to appear with
```python
self.cpp_info.set_property("cmake_find_mode", "none")
```
Is this not the correct way of preventing the generation of *only* the `json2cppConfig.cmake` file?

**Comment by hobbeshunter** [comment]
The problem seems to occur as soon as I set `run=True` and `build=True` on json2cpp in the test package. But without them the executable cannot be found. Is it possible to override this behavior? `headers=True` doesn't seem to work.

**Comment by memsharded** [comment]
Oh, I think I understand better what you mean.
The solution is to add 2 different requirements, like:
```python
def requirements(self):
     self.requires("valijson/version", transitive_headers=True)

def build_requirements(self):
     self.tool_requires("valijson/version")  # equivalent to build=True
```

Because in the general case there are indeed 2 different requirements. For example, if you are cross compiling from Windows to Linux, there will be 2 dependencies:
- The valijson library binary in the "host" context (Linux) that will be used as ``requires`` and used in your application for Linux. It will take the headers from here
- The valijson executable binary in the "build" context (Windows) that will be used as ``tool_requires``, and only used at build time. The library and headers from this package will not be used in the Linux application, of course.

Please try this approach, and let me know.

**Comment by hobbeshunter** [comment]
Thanks that did the job. (I assume you meant json2cpp and not valijson)

Big thanks for conan in general. Makes my live  a lot easier.

**Comment by memsharded** [comment]
> Thanks that did the job. (I assume you meant json2cpp and not valijson)

Yes, sorry, I mixed the two while writing the response.

> Big thanks for conan in general. Makes my live a lot easier.

Many thanks for telling and for your feedback! It makes us very happy to learn that Conan helps making things a bit easier for our users. I checked your profile 😄 , and as a former roboticist myself, and working in C++ in robotics as the initial pain that finally resulted in Conan, it is very rewarding to see Conan being very used in the robotics domain!
